aws_requests_auth/boto_utils.py

```diff
@@ -4,20 +4,50 @@
 aws-requests-auth package.
 """
 
-import boto3
+from botocore.session import Session
 
+from .aws_auth import AWSRequestsAuth
 
-def get_credentials():
+
+def get_credentials(credentials_obj=None):
     """
     Interacts with boto to retrieve AWS credentials, and returns a dictionary of
-    kwargs to be used in AWSRequestsAuth. boto automatically pulls AWS credentials from 
+    kwargs to be used in AWSRequestsAuth. boto automatically pulls AWS credentials from
     a variety of sources including but not limited to credentials files and IAM role.
-    AWS credentials are pulled in the order listed here: 
+    AWS credentials are pulled in the order listed here:
     http://boto3.readthedocs.io/en/latest/guide/configuration.html#configuring-credentials
     """
-    credentials_obj = boto3.Session().get_credentials()
+    if credentials_obj is None:
+        credentials_obj = Session().get_credentials()
+    # use get_frozen_credentials to avoid the race condition where one or more
+    # properties may be refreshed and the other(s) not refreshed
+    frozen_credentials = credentials_obj.get_frozen_credentials()
     return {
-        'aws_access_key': credentials_obj.access_key,
-        'aws_secret_access_key': credentials_obj.secret_key,
-        'aws_token': credentials_obj.token,
+        'aws_access_key': frozen_credentials.access_key,
+        'aws_secret_access_key': frozen_credentials.secret_key,
+        'aws_token': frozen_credentials.token,
     }
+
+
+class BotoAWSRequestsAuth(AWSRequestsAuth):
+
+    def __init__(self, aws_host, aws_region, aws_service):
+        """
+        Example usage for talking to an AWS Elasticsearch Service:
+
+        BotoAWSRequestsAuth(aws_host='search-service-foobar.us-east-1.es.amazonaws.com',
+                            aws_region='us-east-1',
+                            aws_service='es')
+
+        The aws_access_key, aws_secret_access_key, and aws_token are discovered
+        automatically from the environment, in the order described here:
+        http://boto3.readthedocs.io/en/latest/guide/configuration.html#configuring-credentials
+        """
+        super(BotoAWSRequestsAuth, self).__init__(None, None, aws_host, aws_region, aws_service)
+        self._refreshable_credentials = Session().get_credentials()
+
+    def __call__(self, r):
+        # update credentials immediately before each call in case they expired
+        for k, v in get_credentials(self._refreshable_credentials).items():
+            setattr(self, k, v)
```

Comment:
I'm happy to get this in as-is, but I wanted to point out a potential thread safety issue:

- we are in the middle of this `for` loop
- a different thread takes over 
- we've only rotated in new values for, say, `aws_access_key`, `aws_secret_access_key` -- but `aws_token` is still an old "bad" value

One other similar piece of code I've seen is [Yelp's elastalert which has a slightly different implementation of a "refreshable AWS Requests Auth"](https://github.com/Yelp/elastalert/blob/48cb4727ed9189b4340e995b0cc87ea5a296efb9/elastalert/auth.py#L7-L36). I'm not super certain if that implementation is thread safe, but thought it was worth pointing out.

Let me know what you're thinking

Reply:
Good find! The Yelp implementation will work, too, but I think it's significantly less efficient insofar as it will fetch new credentials for *every* HTTP request. Boto is smarter than this, and will cache credentials locally in the session / credentials object until a refresh is due. By storing the Boto credentials object in our `AWSRequestsAuth` instance, we can take advantage of that logic.

I don't think there's a thread safety issue in the this code, because `get_credentials()` has been modified to use the thread-safe `get_frozen_credentials()`, which is guaranteed to return a valid set of credentials (and in no event should the `dict` returned by `get_credentials()` mutate during iteration).  IIRC these credentials are good for at least 5 minutes even after new credentials are available, so for our purposes that's plenty of time to execute an HTTP request and return.

Reply:
ah -- yeah I agree the `get_frozen_credentials()` and the `dict` from `get_credentials` should be stable. 

This is bending my brain a bit, but the scenario I'm thinking about is having two threads running this `for` loop and setting attrs simultaneously: 

- thread 1 and thread 2 are making requests using the same instance of `BotoAWSRequestsAuth` let's call it `boto_aws_auth`
- thread 1 hits `boto_aws_auth.__call__()` and has iterated over and set `boto_aws_auth.aws_access_key` and `boto_aws_auth.aws_secret_access_key` 
- thread 2 takes over 
- thread 2 also hits `boto_aws_auth.__call__()`
- let's say, at this moment, the underlying credentials change 
- thread 2 sets `boto_aws_auth.aws_access_key` and `boto_aws_auth.aws_secret_access_key`to  brand new values distinct from what thread 1 had set before
- thread 1 takes over again
- thread 1 finishes up its for loop and sets `boto_aws_auth.aws_token` to the value that was originally frozen out for thread 1
- thread 1 tries to run its http request
- woops, `boto_aws_auth.aws_access_key` and `boto_aws_auth.aws_secret_access_key` do not match the values originally returned by `get_credentials()` and the request ultimately fails

Bringing this up only so that we can add a note to the README indicating (either way) whether or 
not instances of `BotoAWSRequestsAuth` are safe to use across threads -- thanks for chatting about it!

Reply:
Yes of course -- you're entirely correct. What would you think if passing the credentials into `__call__` instead, and then the parent `__call__` could fall back to the object attributes if no credentials are passed?